src/lib/utils/ct_utils.h

```diff
@@ -173,6 +173,63 @@ class Choice final {
       uint32_t m_value;
 };
 
+/**
+* A CT::Option<T> is either a valid T (Some), or a default T (None)
+*
+* T must be default construtible
+*/
+template <typename T>
+class Option {
+   public:
+      constexpr Option(T v, Choice valid) : m_is_some(valid), m_value(v) {}
+
+      constexpr Option(T v) : m_is_some(Choice::yes()), m_value(v) {}
+
+      constexpr Option() : m_is_some(Choice::no()), m_value(T()) {}
+
+      /// Return if this Option is Some
+      constexpr Choice is_some() const { return m_is_some; }
+
+      /// Return if this Option is None
+      constexpr Choice is_none() const { return !m_is_some; }
+
+      /**
+      * Apply a function to the inner value and return a new Option
+      * which contains that value. This is constant time only if `f` is.
+      */
+      template <typename F>
+      constexpr auto map(F f) const {
+         typedef typename std::invoke_result<F, T>::type U;
+         return Option<U>(f(m_value), m_is_some);
+      }
```

Comment:
```suggestion
      /**
      * Apply a function to the inner value and return a new Option
      * which contains that value. This is constant time only if @p f is.
      *
      * @note The function will always be called, even if the Option is None. It
      *       must be prepared to handle any possible state of T.
      */
      template <std::invocable<const T&> F>
      constexpr auto map(F f) const -> Option<std::invoke_result_t<F, const T&>> {
         return {f(m_value), m_is_some};
      }
```

Reply:
It might make sense to look into the other overloads of [`std::optional::transform`](https://en.cppreference.com/w/cpp/utility/optional/transform). For instance, if `this` is an rvalue-ref we could `std::move(*this)` into the function.

Reply:
Another thing to consider: If `m_is_some` is indicating "None", the resulting `Option<U>` will be "None" but its "value" won't be a default-initialized `U` but whatever came back from `f()`. That may or may not be problematic, but it is probably worth documenting that `m_is_some == None` does not implicate that `m_value` is default inited.